src/simulations/collect.jl

```diff
@@ -164,6 +162,157 @@ function run!(
 end
 
+"""
+    offline_run!(model, agent_step! [, model_step!], n::Integer; kwargs...)
+    offline_run!(model, agent_step!, model_step!, n::Function; kwargs...)
+
+Do the same as [`run`](@ref), but instead of collecting the whole run into an in-memory 
+dataframe, write the output to a file after collecting data `writing_interval` times and 
+empty the dataframe after each write.
+Useful when the amount of collected data is expected to exceed the memory available
+during execution.
+
+## Keywords
+* `backend=:csv` : backend to use for writing data.
+  Currently supported backends: `:csv`, `:arrow`
+* `adata_filename="adata.\$backend"` : a file to write agent data on.
+  Appends to the file if it already exists, otherwise creates the file.
+* `mdata_filename="mdata.\$backend"`: a file to write the model data on.
+  Appends to the file if it already exists, otherwise creates the file.
+* `writing_interval=1` : write to file every `writing_interval` times data collection
+  is triggered. If the `when` keyword is not set, this corresponds to writing to file
+  every `writing_interval` steps; otherwise, the data will be written every
+  `writing_interval` times the `when` condition is satisfied
+  (the same applies to `when_model`).
+"""
+function offline_run! end
+
+offline_run!(model::ABM, agent_step!, n::Int = 1; kwargs...) =
+    offline_run!(model::ABM, agent_step!, dummystep, n; kwargs...)
+
+function offline_run!(model, agent_step!, model_step!, n;
+        when = true,
+        when_model = when,
+        mdata = nothing,
+        adata = nothing,
+        obtainer = identity,
+        agents_first = true,
+        showprogress = false,
+        backend::Symbol = :csv,
+        adata_filename = "adata.$backend",
+        mdata_filename = "mdata.$backend",
+        writing_interval = 1,
+    )
+    df_agent = init_agent_dataframe(model, adata)
+    df_model = init_model_dataframe(model, mdata)
+    if n isa Integer
+        if when == true
+            for c in eachcol(df_agent)
+                sizehint!(c, n)
+            end
+        end
+        if when_model == true
+            for c in eachcol(df_model)
+                sizehint!(c, n)
+            end
+        end
+    end
+
+    run_and_write!(model, agent_step!, model_step!, df_agent, df_model, n;
+        when, when_model,
+        mdata, adata,
+        obtainer, agents_first,
+        showprogress,
+        backend, adata_filename, mdata_filename, writing_interval
+    )
+end
+
+function run_and_write!(model, agent_step!, model_step!, df_agent, df_model, n;
+    when, when_model,
+    mdata, adata,
+    obtainer, agents_first,
+    showprogress,
+    backend, adata_filename, mdata_filename, writing_interval
+)
+    s = 0
+    p = if typeof(n) <: Int
+        ProgressMeter.Progress(n; enabled=showprogress, desc="run! progress: ")
+    else
+        ProgressMeter.ProgressUnknown(desc="run! steps done: ", enabled=showprogress)
+    end
+
+    writer = get_writer(backend)
```

Comment:
Sorry, I wasn't clear with respect to the function barrier. The function barrier has to come _after_ this `writer` object has been initialized, as its initialization is type unstable due to using strings to specify the type of writer.

Reply:
change happened here @fbanning ab617f687cd1910d3095ed8d8f01b20a7c9d4174
just moving the writer initialization back into `offline_run!` and passing the initialized `writer` to `run_and_write!` should solve it

Reply:
Ah, didn't get that part with the function barrier, sorry. I've moved the writer initialisation back into `offline_run!`.